include/jemalloc/internal/ph.h

```diff
@@ -318,36 +318,37 @@ ph_insert(ph_t *ph, void *phn, size_t offset, ph_cmp_t cmp) {
 	 */
 	if (ph->root == NULL) {
 		ph->root = phn;
-	} else {
-		/*
-		 * As a special case, check to see if we can replace the root.
-		 * This is practically common in some important cases, and lets
-		 * us defer some insertions (hopefully, until the point where
-		 * some of the items in the aux list have been removed, savings
-		 * us from linking them at all).
-		 */
-		if (cmp(phn, ph->root) < 0) {
-			phn_lchild_set(phn, ph->root, offset);
-			phn_prev_set(ph->root, phn, offset);
-			ph->root = phn;
-			ph->auxcount = 0;
-			return;
-		}
-		ph->auxcount++;
-		phn_next_set(phn, phn_next_get(ph->root, offset), offset);
-		if (phn_next_get(ph->root, offset) != NULL) {
-			phn_prev_set(phn_next_get(ph->root, offset), phn,
-			    offset);
-		}
-		phn_prev_set(phn, ph->root, offset);
-		phn_next_set(ph->root, phn, offset);
+		return;
 	}
-	if (ph->auxcount > 1) {
-		unsigned nmerges = ffs_zu(ph->auxcount - 1);
-		bool done = false;
-		for (unsigned i = 0; i < nmerges && !done; i++) {
-			done = ph_try_aux_merge_pair(ph, offset, cmp);
-		}
+
+	/*
+	 * As a special case, check to see if we can replace the root.
+	 * This is practically common in some important cases, and lets
+	 * us defer some insertions (hopefully, until the point where
+	 * some of the items in the aux list have been removed, savings
+	 * us from linking them at all).
+	 */
+	if (cmp(phn, ph->root) < 0) {
+		phn_lchild_set(phn, ph->root, offset);
+		phn_prev_set(ph->root, phn, offset);
+		ph->root = phn;
+		ph->auxcount = 0;
+		return;
+	}
+
+	phn_next_set(phn, phn_next_get(ph->root, offset), offset);
+	if (phn_next_get(ph->root, offset) != NULL) {
+		phn_prev_set(phn_next_get(ph->root, offset), phn,
+		    offset);
+	}
+	phn_prev_set(phn, ph->root, offset);
+	phn_next_set(ph->root, phn, offset);
+
+	ph->auxcount++;
+	unsigned nmerges = ffs_zu(ph->auxcount);
```

Comment:
why `ph->auxcount - 1` is changed to `ph->auxcount`? Without the if condition, `ph_try_aux_merge_pair` will be executed even when `auxcount = 1`, which will cause a regression in the performance.

Reply:
When `auxcount = 1` then it has no trailing zero and nothing happens. The only thing the previous check obviates is computing `ffs_zu` for `auxcount = 1` but that's one CPU instruction, but the check itself is more than one instruction, so if anything, this is doing less work now.

Now, you might argue that this will do work when `auxcount = 2` when previously no work was done, and yes, you'd be right. This work is actually necessary, there was an off by one error that left a straggler that never got merged before.

Reply:
I do not think we benefit from the merge being less aggressive here, because the tree gets really unbalanced as one node never acquire children. Work being frontloaded at the insert point is an intentional design decision here, as it is more common to do bulk inserts than it is to do bulk removal, so shifting work from removal to insert improves cache locality.

From my discussion with @davidtgoldblatt on gitter, he also seems to think this was an off by one error in the original code (which he himself wrote).

Reply:
Yeah, I think that this diff is right. We can double-check the perf characteristics with a canary for extra-confidence.